src/arm/dynarec-arm/dynarec-impl.c

```diff
@@ -0,0 +1,1225 @@
+//	printf("%u\n", cpu->nextEvent - cpu->cycles);
+#endif
+
+	if (trace->mode == MODE_ARM) {
+		return;
+	} else {
+		struct ARMDynarecContext ctx = {
+			.code = cpu->dynarec.buffer,
+			.gpr_15 = trace->start + 1 * WORD_SIZE_THUMB,
+			.gpr_15_flushed = false,
+			.gpr_15_invalid = false,
+			.cycles_register_valid = false,
+			.cycles = 0,
+			.prefetch_flushed = false,
+			.is_nzcv_in_host_nzcv = false,
+			.is_reglist_save_pushed = false,
+			.reg_cache_state = REG_CACHE_LOADED,
+		};
+		LOAD_16(ctx.prefetch[0], (trace->start + 0 * WORD_SIZE_THUMB) & cpu->memory.activeMask, cpu->memory.activeRegion);
+		LOAD_16(ctx.prefetch[1], (trace->start + 1 * WORD_SIZE_THUMB) & cpu->memory.activeMask, cpu->memory.activeRegion);
+
+		trace->entry = (void*) ctx.code;
+
+		bool continue_compilation = true;
+		while (continue_compilation) {
+			// emit: if (cpu->cycles >= cpu->nextEvent) return;
+			assert(!ctx.is_reglist_save_pushed && !ctx.gpr_15_invalid);
+			flushNZCVToScratch(&ctx);
+			assertNoAssignedRegs(&ctx);
+			if (!ctx.cycles_register_valid) {
+				EMIT(&ctx, LDRI, AL, REG_SCRATCH0, REG_ARMCore, offsetof(struct ARMCore, cycles));
+				EMIT(&ctx, LDRI, AL, REG_SCRATCH1, REG_ARMCore, offsetof(struct ARMCore, nextEvent));
+				ctx.cycles++; //< there is the +1.
+				// REG_CYCLES contains the expression cycles - nextEvent + 1
+				// This allows us to avoid an unnecessary CMP instruction.
+				EMIT(&ctx, SUB, AL, REG_CYCLES, REG_SCRATCH0, REG_SCRATCH1);
+				ctx.cycles_register_valid = true;
+			}
+			EMIT(&ctx, ADDSI, AL, REG_CYCLES, REG_CYCLES, ctx.cycles);
+			ctx.cycles = 0;
+			EMIT_IMM(&ctx, PL, REG_SCRATCH1, ctx.gpr_15);
+			EMIT(&ctx, B, PL, ctx.code, selectCycleExit(cpu, &ctx));
+
+			// ThumbStep
+			uint32_t opcode = ctx.prefetch[0];
+			ctx.prefetch[0] = ctx.prefetch[1];
+			ctx.gpr_15 += WORD_SIZE_THUMB;
+			LOAD_16(ctx.prefetch[1], ctx.gpr_15 & cpu->memory.activeMask, cpu->memory.activeRegion);
+			ctx.prefetch_flushed = false;
+			ctx.gpr_15_flushed = false;
+			ThumbCompiler instruction = _thumbCompilerTable[opcode >> 6];
+			continue_compilation = instruction(cpu, &ctx, opcode);
+		}
+		flushCycles(&ctx);
+		EMIT(&ctx, B, AL, ctx.code, selectEpilogue(cpu, &ctx));
+
+		ARMDynarecPerformPatching(cpu, trace);
+
+		__clear_cache(trace->entry, ctx.code);
+		cpu->dynarec.buffer = ctx.code;
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define THUMB_PREFETCH_CYCLES (1 + cpu->memory.activeSeqCycles16)
+
+#define THUMB_LOAD_POST_BODY \
+	currentCycles += cpu->memory.activeNonseqCycles16 - cpu->memory.activeSeqCycles16;
+
+#define THUMB_STORE_POST_BODY \
+	currentCycles += cpu->memory.activeNonseqCycles16 - cpu->memory.activeSeqCycles16;
+
+#define FLUSH_HOST_NZCV \
+	flushNZCVToScratch(ctx);
+
+#define PREPARE_FOR_BL \
+	flushCycles(ctx); \
+	flushNZCVFully(ctx);
+
+#define PUSH_REGLIST_SAVE \
+	do { \
```

Comment:
I don't believe the `do { } while (0)` trick is needed with C99, unless you care about scoping.